### ISSUE TITLE
precomputed forward operator obs_seq.out files and obs_sequence_tool

### ISSUE THREAD

**Issue by criedel40** [NCAR/DART#159]
I am working with obs_seq.out files that contain precomputed forward operators. I was hoping to use the obs_sequence_tool program to combine and strip out observations along with their precomputed forward operator values. I noticed in obs_sequence_tool.f90 code there is a 'write_external_FOs' namelist option. However, it is only defined and read in from the namelist. I can not find anymore references to write_external_FOs in the source code. Furthermore, set_obs_def_write_external_FO (the module used to write out obs_seq.out files with precomputed forward operators) is being set from the obs_def_mod in the source code but there is no other reference to it throughout the code either. So it appears the code was going to have this functionality but not finished? Currently when I run obs_sequence_tool with write_external_FOs = .true. I only get the observation values in the newly created obs_seq.out file. Example is located: /glade/p/univ/uokl0016/DAtests/OBS_FILES/ATMOS/test_obs_sequence_tool

**Comment by hkershaw-brown** [comment]
discussion moved to dart email.  Chris is taking a shot at the fix.

**Comment by timhoar** [comment]
I am digging in to the logic here ... the crux of the matter is contained in the comment in the `DEFAULT_obs_def_mod.F90`:

> ! obs_def%write_external_FO should only be true for program 
> ! actually WRITING the external data.  When running DART
> ! obs_def%write_external_FO should be false and no metadata will be written
> ! Also want obs_def%write_external_FO to somehow be true when this called from
> ! the obs_sequence_tool program

I do not understand why this is so.  Under what conditions should DART **_not_** write out the precomputed forward operator values? If the observation definition **_has_** precomputed forward operator values ... why would we not want to write them out? Does it have something to do with writing the posteriors?

If we _always_ write out the precomputed values if available, the logic becomes (I think) trivial, and the code becomes much clearer.

**Comment by nancycollins** [comment]
the precomputed values are stored as metadata in the input obs_seq file.  after filter runs the values are moved to the expected value array, along with any other FOs from obs without precomputed values.  they're output as copies along with mean, sd, etc as requested by the namelist (number of members to write out, etc).  at that point the metadata is redundant and will be unused in any further processing and is removed both for space and less-confusion reasons.

**Comment by criedel40** [comment]
Tim, the reason I modified the code to change obs_def%write_external_FO
depending on the flag set in the namelist (write_external_FOs) was
strictly for the most part working with the obs_sequence_tool. My thinking
behind it is what if someone gives me the obs_seq.out files with
pre-computed FOs but I only want the observations. Example: if the file has
temperature observations from AIRS satellite but the precomputed FOs are
from WRF but I am using MPAS. Using the obs_sequence_tool would provide a
way for someone to take the obs_seq.out file and strip out just the
observations if they wanted too. I agree for the most part that people will
want to keep the precomputed FOs for the most part, however, I just
imagined a situation where someone might want to remove them. Also, I
figured that is why the namelist flag was provided in the first place, to
allow the user to decide if they want to keep them or not.
-Chris

**Comment by nancycollins** [comment]
as long as we're looking at this, i might suggest this as a better design for the obs_sequence_tool namelist.

remove namelist item "write_external_FOs" which isn't well named (it's consistent with the internal code but not with how it's named in the obs_kind mod and other user-facing code.)  instead, add a namelist item called:

discard_precomputed_FOs = "obs_typeA", "obs_typeB"    or  "all"

which is consistent with the obs_kind namelist for precomputed FOs.  
the default would be to keep all externally computed FO metadata, passing obs through unchanged.

then user can chose on an obstype by obstype what to do about the values.

n.

**Comment by weather4evr** [comment]
A few thoughts about this PR:

1) Maybe just personal preference, but there may not be the need for the new subroutine set_obs_precomputed_FOs; just use **call set_obs_precomputed_FOs(seq%obs(i)%def,write_external_FOs)** in set_obs_seq_precomputed_FOs

2) Having to loop over all the obs is inefficient to set the flag.  Instead, if the logic from namelist variable "write_external_FOs" can be pushed into subroutine **write_obs_seq** in **obs_sequence_mod.f90** that would be more efficient, because there's already a loop over the obs when calling **write_obs**. The logic could either be passed as another argument to write_obs_seq, or &obs_sequence_nml (read in **obs_sequence_mod.f90**) could be used instead.

**Comment by timhoar** [comment]
I agree that looping over the obs is inefficient, and am working on an alternate solution. I am trying to take into account most of the comments. I am not going to implement the functionality to selectively choose which observation types have their forward operator values written or ignored or ... but will take a step toward implementing the 'all-or-nothing' approach as a step in the right direction.

**Comment by timhoar** [comment]
There is an existing public routine
```subroutine set_obs_def_write_external_FO(obs_def, write_external_FO)
```
in DEFAULT_obs_def_mod.f90 that allows one to specify whether or not the precomputed FO values are written. By specifying it on a 'per obs_def' basis, you can achieve the kind of granularity that is mentioned in Nancy's `discard_precomputed_FOs` logic. While doing it in a loop is (at this point) overkill, it is extensible.

**Comment by nancycollins** [comment]
just an "observation" (pun intended) that looping over the obs once in a tool is the least expensive part of the tool's runtime, so efficiency by itself shouldn't be a concern.  whether the code is clear and useful functionality is provided is a better metric.

**Comment by timhoar** [comment]
So - an alternate solution (which enables a per-obs-type selection as well as an als-or-nothing approach)  is on a new branch "criedel40-write_FOs_seq_tool"
Please check it out and provide suggestions.

I tested it with the following (cice) 
```
&obs_sequence_tool_nml
   filename_seq       = ''
   filename_seq_list  = 'file.list'
   filename_out       = 'obs_seq.fos'
   print_only         = .false.
   gregorian_cal      = .true.
   synonymous_qc_list = 'COSMOS QC','Ameriflux QC','Data QC','WOD QC'
   min_lat              =   80.0
   max_lat              =   90.0
   min_lon              =  260.5
   max_lon              =  270.5
   external_FO_write_list = 'SAT_SEAICE_AGREG_THICKNESS','AIRS_temperature'
   edit_copies       = .true.
   new_copy_metadata = 'observation'
   /
&preprocess_nml
   input_obs_qty_mod_file  = '../../../assimilation_code/modules/observations/DEFAULT_obs_kind_mod.F90'
   output_obs_qty_mod_file = '../../../assimilation_code/modules/observations/obs_kind_mod.f90'
   input_obs_def_mod_file  = '../../../observations/forward_operators/DEFAULT_obs_def_mod.F90'
   output_obs_def_mod_file = '../../../observations/forward_operators/obs_def_mod.f90'
   obs_type_files          = '../../../observations/forward_operators/obs_def_cice_mod.f90',
                             '../../../observations/forward_operators/obs_def_ocean_mod.f90'
                             '../../../observations/forward_operators/obs_def_AIRS_mod.f90'
   quantity_files          = '../../../assimilation_code/modules/observations/seaice_quantities_mod.f90',
                             '../../../assimilation_code/modules/observations/ocean_quantities_mod.f90',
                             '../../../assimilation_code/modules/observations/atmosphere_quantities_mod.f90'
   /
```

with 
```
0[1526] cheyenne4:/<6>cice/work > cat file.list 
/glade/p/univ/uokl0016/DAtests/OBS_FILES/ATMOS/obs_seq.out_2013082618
/glade/p/cisl/dares/Observations/WOD13/201308/obs_seq.0Z.20130826
```

If this looks good, I can merge it onto master which will close the associated pull request.

**Comment by nancycollins** [comment]
i mentioned this in an email, but i'll also reply here in case anyone else is interested.  i can't see changes to files from an "issue" so i suggest this be made into a separate pull request if the changes are extensive compared to chris' original pull request.  (or if this isn't the best solution, i'm open to whatever people think is the right process.)

**Comment by weather4evr** [comment]
I took a look at the code and I couldn't find anything wrong with it.  The logic for the 'ALL' option is perhaps a little strange, but not too hard to follow and I don't have a better idea.  I didn't try running it myself.

**Comment by timhoar** [comment]
> the precomputed values are stored as metadata in the input obs_seq file. after filter runs the values are moved to the expected value array, along with any other FOs from obs without precomputed values. they're output as copies along with mean, sd, etc as requested by the namelist (number of members to write out, etc). at that point the metadata is redundant and will be unused in any further processing and is not written both for space and less-confusion reasons.

So .... as I understand it, the precompute FO values are intended to appear as the `prior ensemble member      N` copy.
_**This is not the case**_ on the master branch ... I am testing the GSI test data observations with the bgrid model.

**Comment by nancycollins** [comment]
do you have 'use precomputed FOs' set for that obs type in the obs_kind namelist, in addition to being an the 'assimilate these obs types'?  if so, that is where they should end up, 1 for 1, in the obs_seq.final file.  if this obs type isn't in that use namelist then the normal FO is called like all the other "normal" obs.

**Comment by timhoar** [comment]
Argh ... emails crossed paths ... 

of course not!   (UTS)

However, is it by design that even 'compute_posterior = .true.' has no effect on observations with prior FO values? The posterior values for anything with a precomputed FO are MISSING_R8.

**Comment by weather4evr** [comment]
I think it's correct that if the FO values are precomputed, posteriors should always be MISSING_R8 regardless of 'compute_posterior'.  It wouldn't make much sense to have the priors be precomputed but the posteriors using DART's forward operators.

**Comment by timhoar** [comment]
OK - I know we generally discourage putting too much interpretation on the posterior values, but I could conceive of a situation where
people were interested in comparing the posteriors from an experiment using the DART forward operators and the precomputed forward operators. Yes, I know the algorithms to get the prior observation values would be different, and yes, the observation-space innovations would be inconsistent ... but my thought was "We're a testbed ... maybe someone would want to do it." Since it's not immediately apparent that the 'compute_posterior' variable is not applicable to observations with precomputed FOs, I can understand where someone might think it is a bug.

**Comment by nancycollins** [comment]
i think we clearly document that you can't compute posteriors for precomputed FOs to fix that.  if you look at differences in prior vs posterior obs values that are computed in two different ways, you can't attribute changes to the assimilation because some unknown part of the difference can be due to the FO code. 

yes, mechanistically it could be done.  should it?  no.

**Comment by hkershaw-brown** [comment]
Chris's bug fix in in the master branch.  I'm leaving this issue open for now since there is a bunch of discussion still relevant to issue https://github.com/NCAR/DART/issues/168
